calico_node/tests/st/config/test_felix_config.py

```diff
@@ -0,0 +1,410 @@
+
+        if direction == "inbound":
+            spec = {'failsafeInboundHostPorts': udp_ports + tcp_ports}
+        elif direction == "outbound":
+            spec = {'failsafeOutboundHostPorts': udp_ports + tcp_ports}
+        else:
+            self.fail("Adding felix config with wrong 'direction' %s", direction)
+
+        self.add_felix_config(name, spec)
+
+    def run_failsafe(self, failsafe, expect_access):
+        for k in failsafe:
+            for port in failsafe[k]:
+                if port != 179:  # port is already open by bird.
+                    self.open_port(self.host1, port, k)
+
+        for k in failsafe:
+            for port in failsafe[k]:
+                self.check_access_retry(expect_access, self.host2, self.host1, port, k)
+
+    @staticmethod
+    def open_port(host, port, protocal="tcp"):
+        protocal_opt = "-u" if protocal == "udp" else ""
+        cmd = "nc -l %s -p %s -e /bin/sh" % (protocal_opt, port)
+        host.execute(cmd, raise_exception_on_failure=True, daemon_mode=True)
+
+    def check_access_retry(self, expect_access, host_src, host_target, port, protocal="tcp"):
+        msg = (" from %s to %s:%s %s, expect access %s" %
+               (host_src.ip, host_target.ip, port, protocal, expect_access))
+        for retry in range(3):
+            passed = self.check_access(expect_access, host_src, host_target, port, protocal)
+            if passed:
+                _log.info("check access success!%s", msg)
+                return
+            else:
+                _log.exception("retry [%s] check access failed!%s", retry, msg)
+                time.sleep(1)
+        self.fail(("check access failed!%s", msg))
+
+    @staticmethod
+    def check_access(expect_access, host_src, host_target, port, protocal="tcp"):
+        protocol_opt = "-u" if protocal == "udp" else ""
+        cmd = "echo \"hostname && exit\" | timeout -t 3 nc %s -w 1 %s %s" % \
+              (protocol_opt, host_target.ip, port)
+        remote_hostname = host_src.execute(cmd, raise_exception_on_failure=False)
+        hostname = host_target.execute("hostname")
+        _log.info("check access from %s to %s:%s %s, result is |%s|, "
+                  "target hostname |%s|, expect access %s",
+                  host_src.ip, host_target.ip, port, protocal,
+                  remote_hostname, hostname, expect_access)
+
+        # If port is 179, we are talking to bird.
+        # If bird returns with a valid string, this means connection is successful.
+        if port == 179 and remote_hostname is not None and len(remote_hostname) > 0:
+            remote_hostname = hostname
+
+        result = (hostname == remote_hostname)
+        return result == expect_access
+
+    def add_policy(self, policy_data):
+        self._apply_resources(policy_data, self.host1)
+
+    def remove_host_endpoint(self):
+        self.delete_all("hostEndpoint")
+
+    @classmethod
+    def delete_all(cls, resource):
```

Comment:
This one might be useful enough to add to `docker_host.py`?

Reply:
delete_all is not a host operation. It is about datastore manipulation.

Reply:
Fair enough.  `test_base.py` then?  Or some other utilities file somewhere?

Reply:
Let me check.

Reply:
`delete_all_resource` is already in `docker_host.py`. I should use that instead.